### PR TITLE
Add indexing support

### PR DESCRIPTION
**Pull request by eyurtsev**
This PR introduces a persistence layer to help with indexing workflows into
vectostores.

The indexing code helps users to:

1. Avoid writing duplicated content into the vectostore
2. Avoid over-writing content if it's unchanged

Importantly, this keeps on working even if the content being written is derived
via a set of transformations from some source content (e.g., indexing children
documents that were derived from parent documents by chunking.)

The two main components are:

1. Persistence layer that keeps track of which keys were updated and when.
   Keeping track of the timestamp of updates, allows to clean up old content
   safely, and with minimal complexity.
2. HashedDocument which is used to hash the contents (including metadata) of
   the documents. We rely on the hashes for identifying duplicates.

The indexing code works with **ANY** document loader. To add transformations
to the documents, users for now can add a custom document loader
that composes an existing loader together with document transformers.

### CODE REVIEW

**Review by hwchase17 — COMMENTED** [review]
dont we already have `langchain/indexes`? having both `langchain/indexes` and `langchain/indexing` sounds too confusing to me - can we just put it in there?

File: libs/langchain/langchain/indexes/_api.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+        )
+
+
+# PUBLIC API
+
+
+class IndexingResult(TypedDict):
+    """Return a detailed a breakdown of the result of the indexing operation."""
+
+    num_added: Optional[int]
```
Comment (baskaryan) [inline libs/langchain/langchain/indexes/_api.py:134]: why Optional (0 seems like a pretty reasonable "null" values)

File: libs/langchain/langchain/indexes/_api.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+        were added, updated, deleted, or skipped.
+    """
+    if delete_mode == "incremental" and source_id_key is None:
+        raise ValueError("Source id key is required when delete mode is incremental.")
+
+    # Check that the Vectorstore has required methods implemented
+    methods = ["delete", "add_documents"]
+
+    for method in methods:
+        if not hasattr(vector_store, method):
```
Comment (baskaryan) [inline libs/langchain/langchain/indexes/_api.py:201]: i think some methods like `delete` are implemented on base VectorStore but raise NotImplemented. don't suppose there's an easy way to check that here as well?

File: libs/langchain/langchain/indexes/_api.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    # Mark when the update started.
+    index_start_dt = record_manager.get_time()
+    num_added = 0
+    num_skipped = 0
+    num_updated = 0
+    num_deleted = 0
+
+    for doc_batch in _batch(batch_size, doc_iterator):
+        hashed_docs = [
+            _HashedDocument(page_content=doc.page_content, metadata=doc.metadata)
```
Comment (baskaryan) [inline libs/langchain/langchain/indexes/_api.py:222]: nit: isn't there a `from_document`

File: libs/langchain/langchain/indexes/_api.py [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+            for doc in doc_batch
+        ]
+
+        source_ids: Sequence[Optional[str]] = [
+            source_id_assigner(doc.to_document()) for doc in hashed_docs
```
Comment (baskaryan) [inline libs/langchain/langchain/indexes/_api.py:227]:
```suggestion
            source_id_assigner(doc) for doc in doc_batch
```

File: libs/langchain/langchain/indexes/_api.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+                    )
+            # source ids cannot be None after for loop above.
+            source_ids = cast(Sequence[str], source_ids)  # type: ignore[assignment]
+
+        exists_batch = record_manager.exists([doc.uid for doc in hashed_docs])
+
+        # Filter out documents that already exist in the record store.
+        uids = []
+        docs_to_index = []
+        for doc, smart_doc, doc_exists in zip(doc_batch, hashed_docs, exists_batch):
```
Comment (baskaryan) [inline libs/langchain/langchain/indexes/_api.py:248]: smart_doc -> hashed_doc?

File: libs/langchain/langchain/indexes/_api.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+
+        # If source IDs are provided, we can do the deletion incrementally!
+        if delete_mode == "incremental":
+            # Get the uids of the documents that were not returned by the loader.
+            # mypy isn't good enough to determine that source ids cannot be None
+            # here due to a check that's happening above, so we check again.
+
+            for source_id in source_ids:
+                if source_id is None:
+                    raise AssertionError("Source ids cannot be None here.")
```
Comment (baskaryan) [inline libs/langchain/langchain/indexes/_api.py:280]: wasn't this already checked

File: libs/langchain/langchain/indexes/_api.py (hunk 1 above)
Comment (eyurtsev) [reply libs/langchain/langchain/indexes/_api.py:134]: good catch

File: libs/langchain/langchain/indexes/_api.py (hunk 2 above)
Comment (eyurtsev) [reply libs/langchain/langchain/indexes/_api.py:201]: yeah i can add a partial check for that

Comment (eyurtsev) [reply libs/langchain/langchain/indexes/_api.py:229]: Added a check here!

**Comment by baskaryan** [comment]
to make docs look nicer would move notebook to `extras/modules/data_connection/index.ipynb` and make header just `# Indexing`

**Comment by eyurtsev** [comment]
updated to use timestamps for high resolution time, updated notebook with explanation

**Review by baskaryan — APPROVED** [review]

**Pull request merged**